vars/deploy.groovy

```diff
@@ -0,0 +1,149 @@
+        ], config)
+
+        def index = 1
+        def deployments = [:]
+        def deploymentType
+        def deployTool
+
+        if (config.cfTargets) {
+
+            deploymentType = DeploymentType.selectFor(CloudPlatform.CLOUD_FOUNDRY, enableZeroDowntimeDeployment).toString()
+            deployTool = script.commonPipelineEnvironment.configuration.isMta ? 'mtaDeployPlugin' : 'cf_native'
+
+            for (int i = 0; i < config.cfTargets.size(); i++) {
+
+                def target = config.cfTargets[i]
+
+                Closure deployment = {
+
+                    utils.unstash(stageName)
+
+                    cloudFoundryDeploy(
+                        script: script,
+                        juStabUtils: utils,
+                        jenkinsUtilsStub: jenkinsUtils,
+                        deployType: deploymentType,
+                        cloudFoundry: target,
+                        mtaPath: script.commonPipelineEnvironment.mtarFilePath,
+                        deployTool: deployTool
+                    )
+
+                    utils.stash(stageName)
+                }
+                runDeployment(deployments, deployment, index, script, stageName)
+                index++
+            }
+            utils.runClosures(deployments)
+        }
+
+        if (config.neoTargets) {
+
+            deploymentType = DeploymentType.selectFor(CloudPlatform.NEO, enableZeroDowntimeDeployment)
+
+            def source
+            if (config.source) {
+                source = config.source
+            } else {
+                def pom = readMavenPom file: 'application/pom.xml'
+                source = "application/target/${pom.getArtifactId()}.${pom.getPackaging()}"
+            }
+
+            for (int i = 0; i < config.neoTargets.size(); i++) {
+
+                def target = config.neoTargets[i]
+
+                Closure deployment = {
+
+                    utils.unstash(stageName)
+
+                    neoDeploy (
+                        script: script,
+                        warAction: deploymentType.toString(),
+                        source: source,
+                        neo: target
+                    )
+
+                    utils.stash(stageName)
+                }
+                runDeployment(deployments, deployment, index, script, stageName)
+                index++
+            }
+            utils.runClosures(deployments)
+        }
+
+        if (!config.cfTargets && !config.neoTargets) {
+            error "Deployment skipped because no targets defined!"
+        }
+    }
+
+}
+
+void runDeployment(deployments, deployment, index, script, stageName) {
+    deployments["Deployment ${index > 1 ? index : ''}"] = {
+        if (env.POD_NAME) {
+            dockerExecuteOnKubernetes(script: script, containerMap: ContainerMap.instance.getMap().get(stageName) ?: [:]) {
+                deployment.run()
+            }
+        } else {
+            node(env.NODE_NAME) {
+                deployment.run()
```

Comment:
We should discuss whether to call deleteDir() as we cannot be sure what is already available in this node.

Reply:
@OliverNocon 
I have seen that we are also not deleting the workspace in piperStageWrapper. 
https://github.com/SAP/jenkins-library/blob/master/vars/piperStageWrapper.groovy#L39

However, in our stash implementation we do: https://github.wdf.sap.corp/MA/cloud-s4-sdk-pipeline-lib/blob/master/vars/unstashFiles.groovy#L6

As far as I know node will create a new workspace, but the workspace can be one which is reused from before. So there is no gerantee which files are already there, right?

We should also discuss whether we should introduce delete files in piperStageWrapper.

Reply:
the `piperStageWrapper` does the deleting at the end of the stage. This should be enough for most cases. Having a an additional `deleteDir()` at the beginning is a bit more bullet proof but not sure if this is really needed.
Anyway, this should not hold up this PR